### PR TITLE
Feature/vault integration

### PR DESCRIPTION
**Pull request by Kidswiss**
With this commit the cluster service account token will be added to
    vault.

Closes: #33

### CODE REVIEW

File: pkg/controller/cluster/cluster_reconcile.go [hunk 1]
```diff
@@ -63,6 +68,21 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	token, err := r.getServiceAccountToken(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = client.SetToken(path.Join(instance.Spec.TenantRef.Name, instance.Name), token, reqLogger)
```
Comment (srueg) [inline pkg/controller/cluster/cluster_reconcile.go:81]: The token should be saved in `<tenantId>/<clusterId>/steward` with `token=<token>`
```suggestion
	err = client.SetToken(path.Join(instance.Spec.TenantRef.Name, instance.Name, "steward"), token, reqLogger)
```

File: pkg/vault/client.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	"os"
+	"path"
+
+	"github.com/banzaicloud/bank-vaults/pkg/sdk/vault"
+	"github.com/go-logr/logr"
+	"github.com/hashicorp/vault/api"
+)
+
+const (
+	tokenName    = "steward token"
```
Comment (srueg) [inline pkg/vault/client.go:13]:
```suggestion
	tokenName    = "token"
```

File: pkg/vault/client.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	k8sTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+	tokenEnvVar  = "VAULT_TOKEN"
```
Comment (srueg) [inline pkg/vault/client.go:15]: We could reuse `api.EnvVaultToken` from the Vault client

File: pkg/vault/client.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	vaultEnvVar  = "VAULT_URL"
```
Comment (srueg) [inline pkg/vault/client.go:16]: We could reuse `api.EnvVaultAddress` from the Vault client

File: pkg/controller/cluster/cluster_reconcile.go [hunk 5]
```diff
@@ -141,3 +161,24 @@ func (r *ReconcileCluster) updateTenantGitRepo(tenant types.NamespacedName, file
...
+	if err != nil {
+		return "", err
+	}
+
+	sortSecrets := helpers.SecretSortList(*secrets)
+
+	sort.Sort(sort.Reverse(sortSecrets))
+
+	for _, secret := range sortSecrets.Items {
+		if secret.Annotations[corev1.ServiceAccountNameKey] == instance.Name {
```
Comment (srueg) [inline pkg/controller/cluster/cluster_reconcile.go:178]: Let's also check for `secret.Type == corev1.SecretTypeServiceAccountToken` and if the secret really has a key `secret.Data["token"]` (if a new secret is created, it doesn't have the token yet)...

File: pkg/controller/cluster/cluster_reconcile_test.go [hunk 6]
```diff
@@ -85,6 +87,17 @@ func TestReconcileCluster_Reconcile(t *testing.T) {
 						GitRepoTemplate: &synv1alpha1.GitRepoTemplate{},
 					},
 				},
+				&corev1.Secret{
```
Comment (srueg) [inline pkg/controller/cluster/cluster_reconcile_test.go:90]: Would be nice to have a test case with multiple secrets, to test the sorting

File: pkg/vault/client.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	secret, err := b.client.RawClient().Logical().Read(queryPath)
+	if err != nil {
+		return err
+	}
+
+	if secret == nil || secret.Data[tokenName] != token {
+
+		log.WithName("vault").Info("secrets don't match, re-applying")
+
+		data := map[string]interface{}{
```
Comment (srueg) [inline pkg/vault/client.go:89]: We should probably reuse the `secret` map here to prevent removing other entries than `token`

Comment (Kidswiss) [reply pkg/vault/client.go]: While looking into this I found quite the bug in the update logic, where I didn't unmarshal the secret correctly so it was updated on every reconcile-loop. That's now fixed.

File: pkg/controller/cluster/cluster_reconcile.go (hunk 5 above)
Comment (Kidswiss) [reply pkg/controller/cluster/cluster_reconcile.go]: This returns now a specific error that gets checked in the reconcile function and then ignored. So we don't trigger unnecessary reconciles while waiting for the token to be populated.

File: pkg/controller/cluster/cluster_reconcile.go [hunk 8]
```diff
@@ -63,6 +68,23 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
 	repoName.Name = instance.Spec.TenantRef.Name
 
+	client, err := vault.NewClient()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	token, err := r.getServiceAccountToken(instance)
+	if err != nil {
+		if err.Error() != "no token available yet" {
```
Comment (srueg) [inline pkg/controller/cluster/cluster_reconcile.go:78]: Please create a proper error instead of string comparison

File: pkg/vault/client.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+		log.WithName("vault").Info("does not yet exist, creating", "name", secretPath)
+		secret = &api.Secret{}
+		secret.Data = vault.NewData(0, map[string]interface{}{
+			tokenName: token,
+		})
+		_, err = b.client.RawClient().Logical().Write(queryPath, secret.Data)
+		return err
+	}
+
+	secretData := secret.Data["data"].(map[string]interface{})
```
Comment (srueg) [inline pkg/vault/client.go:93]: Maybe add a type check?

File: pkg/vault/client_test.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+func testGetHTTPServer(statusCode int, body []byte) *httptest.Server {
+	testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		res.WriteHeader(statusCode)
+		_, _ = res.Write(body)
+	}))
+
+	return testServer
+}
+
+func TestNewClient(t *testing.T) {
```
Comment (srueg) [inline pkg/vault/client_test.go:22]: Maybe add another test for the `SetToken()` create/update path?

File: pkg/controller/cluster/cluster_reconcile.go (hunk 8 above)
Comment (srueg) [reply pkg/controller/cluster/cluster_reconcile.go:78]: Also this should still `return`, right?
Thinking about it, I don't think this should be a special case. It's just an error and should be handled like one...

File: pkg/controller/cluster/cluster_reconcile_test.go [hunk 11]
```diff
@@ -85,6 +85,32 @@ func TestReconcileCluster_Reconcile(t *testing.T) {
...
+						Annotations: map[string]string{
+							corev1.ServiceAccountNameKey: tt.fields.objName,
+						},
+						CreationTimestamp: metav1.Time{Time: time.Now()},
+					},
+					Type: corev1.SecretTypeServiceAccountToken,
+					Data: map[string][]byte{
+						"token": []byte("mysecret1"),
+					},
+				},
```
Comment (srueg) [inline pkg/controller/cluster/cluster_reconcile_test.go:113]: Let's add a test case with a secret and missing `token`.

File: pkg/controller/cluster/cluster_reconcile.go (hunk 8 above)
Comment (Kidswiss) [reply pkg/controller/cluster/cluster_reconcile.go:78]: As discussed we simply skip this case and return the first found secret with a token.

File: pkg/vault/client.go (hunk 9 above)
Comment (Kidswiss) [reply pkg/vault/client.go:93]: As discussed: makes no sense, as the type of the `data` field is defined by the secret engine. If that's not correct we're talking to the wrong engine anyways...

File: pkg/controller/cluster/cluster_reconcile_test.go (hunk 11 above)
Comment (Kidswiss) [reply pkg/controller/cluster/cluster_reconcile_test.go:113]: Added a case where the newest secret doesn't have a token and falls back to the previous one.

File: pkg/vault/client_test.go (hunk 10 above)
Comment (Kidswiss) [reply pkg/vault/client_test.go:23]: Added separate test case

**Pull request merged**